### PR TITLE
Add `ArcSin` bloq

### PR DESCRIPTION
**Pull request by charlesyuan314**
The deccomposition of this bloq is quite tricky (polynomial interpolation followed by some arithmetic and square roots), but for now we start with its Toffoli count as derived from the referenced paper.

### CODE REVIEW

File: qualtran/bloqs/arithmetic/trigonometric.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+from attrs import frozen
+
+from qualtran import Bloq, bloq_example, BloqDocSpec, QFxp, Register, Signature
+from qualtran.bloqs.basic_gates import Toffoli
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+from qualtran.symbolics import is_symbolic, SymbolicInt
+
+
+@frozen
+class ArcSin(Bloq):
```
[the remark this answers is not in the document]
Comment (charlesyuan314) [reply qualtran/bloqs/arithmetic/trigonometric.py:26]: Happy to, but question: I saw [this comment](https://github.com/tanujkhattar/Qualtran/blob/13827a41b5f8b2ae4c02eda30b15c6001c7cadfe/qualtran/_infra/data_types.py#L470) which suggested that classical values for `QFxp` are not correctly propagated through bloqs.

Is that something to watch out for?

[the remark this answers is not in the document]
Comment (tanujkhattar) [reply qualtran/bloqs/arithmetic/trigonometric.py:26]: There's ongoing work to fix this, see https://github.com/quantumlib/Qualtran/pull/1142

You can look at implementation of `AddScaledValIntoPhaseReg` to see how to implement the simulation protocols for bloqs expecting `QFxp` to the best of the ability of the infrastructure we have right now.

[the remark this answers is not in the document]
Comment (charlesyuan314) [reply qualtran/bloqs/arithmetic/trigonometric.py:29]: I added `on_classical_vals` and tests for it.

I'd like to add `my_tensors` in a follow-up PR given its increased complexity.

File: dev_tools/autogenerate-bloqs-notebooks-v2.py [hunk 2]
```diff
@@ -54,17 +54,33 @@
...
+import qualtran.bloqs.block_encoding.tensor_product
+import qualtran.bloqs.block_encoding.unitary
 import qualtran.bloqs.bookkeeping
+import qualtran.bloqs.bookkeeping.allocate
+import qualtran.bloqs.bookkeeping.auto_partition
+import qualtran.bloqs.bookkeeping.cast
+import qualtran.bloqs.bookkeeping.free
+import qualtran.bloqs.bookkeeping.join
+import qualtran.bloqs.bookkeeping.partition
+import qualtran.bloqs.bookkeeping.split
```
Comment (mpharrigan) [inline dev_tools/autogenerate-bloqs-notebooks-v2.py:83]: is this a merge thing or an orthogonal cleanup thing or something else?

File: qualtran/bloqs/arithmetic/trigonometric.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        result: `bitsize`-sized output register.
+
+    References:
+        [Optimizing Quantum Circuits for Arithmetic](https://arxiv.org/abs/1805.12445). Appendix D.
+    """
+
+    bitsize: SymbolicInt
+    num_frac: SymbolicInt
+    num_iters: SymbolicInt = 4  # reference studies 3, 4, or 5 iterations
+    degree: SymbolicInt = 4  # reference studies degree-3, 4, 5, or 6 polynomials
```
Comment (mpharrigan) [inline qualtran/bloqs/arithmetic/trigonometric.py:54]: can you move these comments (which might be missed or re-formatted) to the Args section of the docs? I.e. you can say "The default is 4 and the reference studies 3, 4, or 5 iterations"

File: qualtran/bloqs/arithmetic/trigonometric.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        result ^= int(np.arcsin(x_fxp) * 2**self.bitsize)
+        return {'x': x, 'result': result}
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        n = self.bitsize
+        p = self.bitsize - self.num_frac
+        d = self.degree
+        m = self.num_iters
+        # directly copied from T_arcsin on page 10 of reference
+        ts = (
```
Comment (mpharrigan) [inline qualtran/bloqs/arithmetic/trigonometric.py:91]: just to confirm: this is the number of toffoli gates, not t gates?

File: dev_tools/autogenerate-bloqs-notebooks-v2.py (hunk 2 above)
Comment (charlesyuan314) [reply dev_tools/autogenerate-bloqs-notebooks-v2.py:83]: It is an orthogonal cleanup. I will move it to a separate PR.

Comment (charlesyuan314) [reply dev_tools/autogenerate-bloqs-notebooks-v2.py]: Done in #1195.

**Review by mpharrigan — APPROVED** [review]

**Pull request merged**